pbrd/pbr_vty.c

```diff
@@ -153,325 +128,949 @@ DEFUN_NOSH(no_pbr_map, no_pbr_map_cmd, "no pbr-map PBRMAP [seq (1-700)]",
+ * address family from resolved nexthops or nexthop-groups. All of
+ * those opportunistic address-family determinations are stored in
+ * the "family" field of struct pbr_map_sequence.
+ *
+ * This "family" field value is needed particularly when deleting
+ * a rule piece-by-piece because at the end, the match/set fields
+ * will be empty. Maybe it would be possible to handle this issue
+ * as an internal zebra matter in the future.
+ *
+ * We also attempt to maintain address-family consistency among the
+ * various configured fields in a rule. So far, these fields are
+ * src/dst IP-address match/set values.
+ *
+ * It is probably possible to perform the same address-family check in
+ * the CLI for single nexthops (set nexthop A.B.C.D|X:X::X:X) but the
+ * address-family is not immediately available for nexthop-groups.
+ * In both the single-nexthop and nexthop-group, the NHT resolution code
+ * sets the "family" field of struct pbr_map_sequence asynchronously.
+ *
+ * There isn't currently any flagging of rules that have a consistent
+ * set of src/dst IP-address match/set values but an asynchronously-resolved
+ * nexthop-group that has a different address-family.
+ *
+ * The match/set IP-address handlers below blindly set "family"; it's
+ * probably possible to wrongly set "family" to, e.g., IPv4 this way after
+ * a v6 NHG has been resolved and break rule removal. It's not clear
+ * how to best address this potential issue.
+ */
+static bool pbr_family_consistent(struct pbr_map_sequence *pbrms,
+				  uint8_t family, uint32_t skip_filter_bm,
+				  uint32_t skip_action_bm, const char **msg)
+{
+	uint32_t filter_bm = pbrms->filter_bm & ~skip_filter_bm;
+	uint32_t action_bm = pbrms->action_bm & ~skip_action_bm;
+
+	if (CHECK_FLAG(filter_bm, PBR_FILTER_SRC_IP) &&
+	    (family != pbrms->src->family)) {
+		if (msg)
+			*msg = "match src-ip";
+		return false;
 	}
-
-	return ret;
+	if (CHECK_FLAG(filter_bm, PBR_FILTER_DST_IP) &&
+	    (family != pbrms->dst->family)) {
+		if (msg)
+			*msg = "match dst-ip";
+		return false;
+	}
+	if (CHECK_FLAG(action_bm, PBR_ACTION_SRC_IP) &&
+	    (family != sockunion_family(&pbrms->action_src))) {
+		if (msg)
+			*msg = "set src-ip";
+		return false;
+	}
+	if (CHECK_FLAG(filter_bm, PBR_ACTION_DST_IP) &&
+	    (family != sockunion_family(&pbrms->action_dst))) {
+		if (msg)
+			*msg = "set dst-ip";
+		return false;
+	}
+	return true;
 }
 
-DEFPY(no_pbr_set_table_range, no_pbr_set_table_range_cmd,
-      "no pbr table range [(10000-4294966272)$lb (10000-4294966272)$ub]",
-      NO_STR
-      PBR_STR
-      "Set table ID range\n"
-      "Set table ID range\n"
-      "Lower bound for table ID range\n"
-      "Upper bound for table ID range\n")
+
+/* clang-format off */
+DEFPY  (pbr_map_match_src,
+	pbr_map_match_src_cmd,
+	"[no] match src-ip ![<A.B.C.D/M|X:X::X:X/M>$prefix]",
+	NO_STR
+	"Match the rest of the command\n"
+	"Choose the src ip or ipv6 prefix to use\n"
```

Comment:
Just a wording: `ipv4 or ipv6` maybe? :)

Reply:
Agreed; fixed for match/set src-ip/dst-ip